Modules/pyexpat.c

```diff
@@ -1654,7 +1654,18 @@ static int
 add_error(PyObject *errors_module, PyObject *codes_dict,
           PyObject *rev_codes_dict, const char *name, int value)
 {
-    const char *error_string = XML_ErrorString(value);
+    const char * const error_string_or_null = XML_ErrorString(value);
+
+    /* Be robust towards runtime libexpat being too old to know
+     * all the latest error codes */
+    const char *error_string = error_string_or_null;
```

Comment:
@hartwork 
Out of curiosity, since CPython vendors `libexpat` to be expected version.
Is there a possibility of adding unexpected constants that XML_ErrorString will return null?

Reply:
Hi @corona10 , very valid question! `XML_ErrorString` returns `NULL` whenever the implementation does not know the error code that is being queried. That happens (a) for invalid values and (b) for values from the future, e.g. when calling `XML_ErrorString(XML_ERROR_AMPLIFICATION_LIMIT_BREACH)` with libexpat <2.4.0.

In general, there is (1) the libexpat version used at _build time_ and (2) the expat version used at _runtime_, loaded through dynamic linking. And they can be different. So in my mind there are two questions:

- Is there a chance that someone builds CPython with (vulnerable) libexpat <2.4.0 (and is it okay to fail the build if so)?
- Is there a chance that someone uses CPython binaries with libexpat binaries other than (and older than) the ones that CPython was built against (and is it okay to fail loading module `xml.parsers.expat.errors` if so)?

For status quo of this pull request I decided I would want to err on the side of caution for the case with older runtime. For compile time, it can be argued that libexpat <2.4.0 is vulnerable, and that hence the person compiling against libexpat <2.4.0 needs to get off of their vulnerable copy anyway. But that's my personable take, I'm happy to discuss. We could protect against both or non of these. What do you think?

Reply:
> Is there a chance that someone builds CPython with (vulnerable) libexpat <2.4.0 (and is it okay to fail the build if so)?

My personal opinion is okay to fail because the current implementation does not protect such situations,
if such uses cases were a lot, related issues are already created but it looks not.

But let's hear @vstinner and @tiran 's opinion for both questions, they know more about a lot of related history.

Merry Christmas.

Reply:
> the current implementation does not protect such situations

@corona10 I think I should note that that the current implementation only uses error codes that are so old, it doesn't really have that problem. But adding new codes changes the game and raises these two new questions to be answered one way or another.

Merry Christmas to you and your family.

Reply:
@hartwork 
> I think I should note that that the current implementation only uses error codes that are so old,

Ah I see, in that case, what about adding version checking macro for those constants, or what about detecting such constants by macro and adding constants at compile time?
I didn't check libexpat can use such approaches.

see similar approach: https://github.com/python/cpython/blob/078abb676cf759b1e960f78390b6e80f256f0255/Modules/_gdbmmodule.c#L128

Reply:
@corona10 building pre-processor guards from https://github.com/libexpat/libexpat/blob/9542ad247130801eb226225299d01a17c0582f08/expat/lib/expat.h#L1042-L1044 would be easy, but it would mean that the set of available codes now become "dynamic" in the sense that the user can no longer assume that all those error code constants are available. Maybe that's worse than making the build require libexpat >=2.4.0 (some clean way). What do you think?

Reply:
@hartwork 
> Maybe that's worse than making the build require libexpat >=2.4.0 (some clean way). 

If you worry about custom libexpat build, IMHO dynamic constant adding is useful and we can announce it the documentation. 

But I concur with the fixed version requirement, but it will need some discussion with other core devs.
You can use a mailing list or Discourse (https://discuss.python.org/c/core-dev/).
Since you are libexpat module maintainer, I think that your opinion will be persuasive.